app/ui/src/app/integration/edit-page/step-configure/step-configure.component.ts

```diff
@@ -138,16 +135,38 @@ export class IntegrationStepConfigureComponent extends FlowPage implements OnIni
     }
     const prevStep = this.currentFlow.getPreviousStepWithDataShape(this.position);
     const nextStep = this.currentFlow.getSubsequentStepWithDataShape(this.position);
-    this.currentFlow.fetchOutputDataShapeFor(prevStep)
-      .then(inDataShape => {
-        this.inputDataShape = inDataShape;
-        this.currentFlow.fetchInputDataShapeFor(nextStep)
-          .then(outDataShape => {
-            this.outputDataShape = outDataShape;
-          })
-          .catch(response => this.handleDataShapeError(nextStep, response));
+
+    // we now have previous and next steps that have the data shapes defined
+    // now we need to determine if there is a mapping step inbetween
+    // mapping step doesn't specify shape but it changes it to the input shape
+    // of the step after it, in that case we need to look at the input sep of
+    // the next step, otherwise we need to look at the output step of the
+    // previous step
+
+    const prevIdx = this.currentFlow.steps.indexOf(prevStep);
+    const nextIdx = this.currentFlow.steps.indexOf(nextStep);
```

Comment:
So I really need to hide that `steps` attribute from currentFlow :-)

You know, maybe a better solution is to just add an `action` property to a data mapper step and just set the input/output data shape on it from the data mapper host component, then this code simply becomes `const step = currentFlow.getPreviousStepWithDataShape(this.position);` and the filter always sees the output data shape?

I know the plan is to add data shapes to mapper steps, right @igarashitm?

Mucking around with the array index scares me basically :-)  It feels like just doing `currentFlow.getPreviousSteps()`, checking if any of them has a mapper and then getting the previous or subsequent step's data shape is safer.

Reply:
We definitely need to add action and input&output data shape on datamapper step (#1394). This is required to achieve visual indicator of the data shapes for each steps (#700).

Reply:
Ok cool.  Since that's still outstanding I'm fine for merging this, one the datamapper step has data shapes this will greatly simplify this.